src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs

```diff
@@ -18,6 +18,14 @@ namespace System.Runtime.CompilerServices
         internal StringHandleOnStack([NotNull] ref string? s)
         {
             _ptr = Unsafe.AsPointer(ref s);
+
+            // The StringHandleOnStack type itself doesn't guarantee that 's' is
+            // non-null. But the typical usage pattern is that this handle is
+            // passed to a QCall routine, and that QCall routine will ensure that
+            // the captured reference is non-null. So even though the [NonNull]
+            // annotation here is a lie we'll keep it around for convenience.
+
+            NullabilityHelpers.SuppressNonNullAssignmentWarning(ref s);
```

Comment:
This isn't necessary.  You can just change the:
```C#
_ptr = Unsafe.AsPointer(ref s);
```
to be:
```C#
_ptr = Unsafe.AsPointer(ref s!);
```

But, as commented earlier, I think the right solution instead is to remove the [NotNull].

Reply:
(Call sites can then use `!` as is appropriate, e.g. `string s = null!` or `ref s!`.)